### PR TITLE
[opt](nereids) range inference convert isnull to EmptyValue 

### PR DESCRIPTION
**Pull request by yujun777**
### What problem does this PR solve?

Issue Number: close #xxx

Related PR: #xxx

Problem Summary:

for sql 

```
a > 10 and a < 20 or a > 30 and a < 40 or  a > 60 and a < 50
```

AddMinMax should opt it as: 

```
(a < 20 or a > 30 or a is null and null) and a > 10 and a < 40
```

But it fail, because the SimplifyRange will opt it as 

```
a > 10 and a < 20 or a > 30 and a < 40 or  a is null and null
```

since the 3rd is `a is null and null`,  it couldn't get a's range and suppose a's range is [-00, +00], then the whole expression,  it will get a's range is [-00, +00], to fix this problem,  RangeInference need to convert `a is null and null` to EmptyValue instead of UnknownValue.

### Release note

None

### Check List (For Author)

- Test 
    - [ ] Regression test
    - [x] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

**Comment by yujun777** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java [hunk 1]
```diff
@@ -130,10 +132,22 @@ private ValueDesc simplify(ExpressionRewriteContext context,
             Expression originExpr, List<Expression> predicates,
             BinaryOperator<ValueDesc> op, boolean isAnd) {
 
+        boolean convertIsNullToEmptyValue = isAnd && predicates.stream().anyMatch(expr -> expr instanceof NullLiteral);
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:135]: it is better to check if predicates contains "isNull" and "null" literal.
if match, return EmtpyValue directly, and skip the for loop in line138-150

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java [hunk 2]
```diff
@@ -469,6 +486,9 @@ public ValueDesc union(ValueDesc other) {
 
         @Override
         public ValueDesc intersect(ValueDesc other) {
+            if (other instanceof EmptyValue) {
+                return other.intersect(this);
```
Comment (englefly) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:490]: why not return EmptyValue?

**Review by starocean999 — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java [hunk 3]
```diff
@@ -326,7 +326,10 @@ private Map<Expression, MinMaxValue> getExprMinMaxValues(UnknownValue valueDesc)
         for (int i = 1; i < sourceValues.size(); i++) {
             // process in sourceValues[i]
             Map<Expression, MinMaxValue> minMaxValues = getExprMinMaxValues(sourceValues.get(i));
-            for (Map.Entry<Expression, MinMaxValue> entry : minMaxValues.entrySet()) {
+            List<Map.Entry<Expression, MinMaxValue>> minMaxValueList = minMaxValues.entrySet().stream()
+                    .sorted((a, b) -> Integer.compare(a.getValue().exprOrderIndex, b.getValue().exprOrderIndex))
+                    .collect(Collectors.toList());
+            for (Map.Entry<Expression, MinMaxValue> entry : minMaxValueList) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java:332]: add comment to explain why need sort by expr id

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java [hunk 4]
```diff
@@ -461,6 +475,9 @@ public boolean isAnd() {
 
         @Override
         public ValueDesc union(ValueDesc other) {
+            if (other instanceof EmptyValue) {
+                return other.union(this);
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:480]: add some comment to explain this

**Comment by yujun777** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java (hunk 1 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:135]: this pr only solved user case. other pr will fix this.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java (hunk 2 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:494]: its final result indeed return EmptyValue. Bug for unify function purpose,  when RangeValue/DiscreteValue/UnknownValue interset with an EmptyValue, call EmptyValue.interset(this) => EmptyValue

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java (hunk 3 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java:339]: during mergging sourceValues[i] into result,  we want to keep the relative order of values of sourceValues[i] if they don't  in result.  for example if a's exprOrderIndex < b's exprOrderIndex,  then in result, a's new exprOrderIndex also < b's new exprOrderIndex.

also add this comment in code.

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java (hunk 4 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:482]: for a given reference,   EmptyValue(reference) or UnknownValue(reference),  then it should always evaluate to UnknowValue(referece)

**Review by starocean999 — APPROVED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java (hunk 3 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java:339]: add these comment into code please

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java (hunk 4 above)
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:482]: add these comment into code please

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java (hunk 3 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/AddMinMax.java:339]: had add

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java (hunk 4 above)
Comment (yujun777) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/expression/rules/RangeInference.java:482]: had add

**Review by morrySnow — APPROVED** [review]

**Pull request merged**